File: ga_model_2P1Z_v8.m
function varargout=ga_model_2P1Z_v8(Nsupply,C_nut,varargin)
%% GA_MODEL_2P1Z_FROMPSUPPLY
% Default parameters are based on Messié & Chavez (2017).
% 
% Use:
% [output=]ga_model_2P1Z_*(Nsupply,varargin)
%
% output is a structure containing 
%	.time, .Nsupply, .P_small, .P_big, .Z_small, .Z_big, 
%	.Nnew, .Nreg, .Chl, .PP, .u_small, .u_big, .g_small, .g_big, 
%	.units, .attributs
%
% Required input:
% 	Nsupply expressed in mmolC/m3/d, all units are carbon-based. Note - Nnew and Nreg represent new and regenerated PO4rients, respectively
%		(NO3 and NH4 as a simplification) but are termed "Nnew" and "Nreg" to limit confusion with the unit, since they are expressed in carbon.
% 	Nsupply can be either a number, corresponding to the rate observed during upw_duration (default 1 day following Messié & Chavez 2017)
%					or a vector (then time needs to be given), that provides Nsupply as a function of time along a current trajectory,
%					for instance (useful to take Ekman pumping into account)
%
% Optional inputs:
% 'nbdays_advec'	number of days during which the model is run
% 'dt'				time step
% 'time'			can replace nbdays_advec and dt (required if Nsupply is a vector)
% 'upw_duration'	number of days during which Nsupply happens (default 1 day, not used if Nsupply is a vector)
% 'plot'			displays the plankton model outputs as a function of time
%
% Monique Messié, 2021 for public version
% Reference: Messié, M., & Chavez, F. P. (2017). PO4rient supply, surface currents, and plankton dynamics predict zooplankton hotspots 
%					in coastal upwelling systems. Geophysical Research Letters, 44(17), 8979-8986, https://doi.org/10.1002/2017GL074322
% Differences with Messié and Chavez (2017):
%		all Zsmall excretion is now availabe as regenerated nutrients (ie no export on Zsmall excretion)
%		Zbig grazing formulation is different with the half-saturation constant applying to Z_small+P_big in both cases


%% -------------- Default parameters (see Messié & Chavez, 2017 suppl inf)


default_parameters={...
'umax_small',1.9872,'umax_big',2.7648,'gmax1',1.1126,'gmax2',1.4226,...% maximum growth and grazing rates (d^{-1}) (Baklouti et al., 2021)
'cChl_small',200,'cChl_big',50,...								% C:Chl ratios for Psmall and Pbig (only used to calculate Chl) (Lazzari et al., 2012)
'kP_small',13.3,...											% half-saturation constant for Psmall on PO4 (mmolC m^{-3}) (Pulido-Villena et al., 2021) )
'kP_big',15.3,...												% half-saturation constant for Pbig on PO4 (mmolC m^{-3}) (Pulido-Villena et al., 2021)
'kG',5,...										% half-saturation constant for Z on P (mmolC m^{-3}) (Auger et al., 2011)
'mP',0,...														% Pbig mortality rate (default 0 ie no Pbig sinking) (d^{-1}) (Auger et al., 2011)
'mZ',0.005,...											% Zbig quadratic mortality rate (mmolC^{-1} m^{3} d^{-1})
'eZ',0.1,...													% zoo excretion fraction (Zsmall and Zbig) (d^{-1}) (Baklouti et al., 2021)
'epsilon',0.75 ,...												% fraction of Z excretion that is available as regenerated PO4rients
'P_small_ini',0.3,'P_big_ini',0.1,'Z_ini',0.7};	% initial biomass (mmolC m^{-3}) (In situ for P_small_ini with method from Marrec et al., 2018 and Tzortzis et al., 2021 & further)

[arg,flag]=ga_read_varargin(varargin,[{'nbdays_advec',20,'dt',0.2,'time',[]},default_parameters],{'plot'});
if length(Nsupply)>1 && isempty(arg.time), error('Give time if Nsupply is a vector'), end


%% -------------- Time

if isempty(arg.time), time=(0:arg.dt:arg.nbdays_advec)'; 
else, time=arg.time(:); arg.dt=time(2)-time(1); 
end
nb_time=length(time); 



%% -------------- Nsupply

if length(Nsupply)==1
	Nsupply=zeros(nb_time,1);
end



%% -------------- Initial conditions

%Biomasses
PO4=time*NaN; 				% PO4rients (PO4 expressed in carbon) (mmolC m^{-3})
P_small=time*NaN; 			% small phyto biomass (mmolC m^{-3})
P_big=time*NaN; 			% large phyto biomass (mmolC m^{-3})
Z=time*NaN; 			% small zoo biomass (mmolC m^{-3})
u_big=time*NaN; 	        % P_big growth rate (Nnew-limited) (d^{-1})
u_small=time*NaN; 			% P_small growth rate (Nreg-limited) (d^{-1})
g=time*NaN; 			% Z_big grazing rate (P_big and Z_small limited) (d^{-1})
PP_big=time*NaN; 			% primary production for P_big (mmolC m^{-3} d^{-1})
PP_small=time*NaN;         % primary production for P_small (mmolC m^{-3} d^{-1})        

%Flux
G1=time*NaN; 			% grazing from Z onto Pbig (mmolC m^{-3} d^{-1})
G2=time*NaN; 			% grazing from Z onto Psmall (mmolC m^{-3} d^{-1})
excretion_Z=time*NaN; 	% Z excretion (mmolC m^{-3} d^{-1})
death_Z=time*NaN; 		% Zbig mortality term (mmolC m^{-3} d^{-1})
regeneration_PO4=time*NaN;	% PO4rient regeneration (mmolC m^{-3} d^{-1})
p_fec=time*NaN;
Export=time*NaN;
Budget=time*NaN;

%Valeurs initiales
P_small(1)=arg.P_small_ini; 
P_big(1)=arg.P_big_ini; 
Z(1)=arg.Z_ini;
PO4(1)=C_nut;
Export(1) = 0.021;
Budget(1) = 0.381;



%% -------------- Loop on time

for t=2:nb_time

	% growth and grazing rates
    u_big=PO4(t-1)/(arg.kP_big+PO4(t-1))*arg.umax_big;
    u_small=PO4(t-1)/(arg.kP_small+PO4(t-1))*arg.umax_small;
  
    g1=P_big(t-1)/(arg.kG+P_small(t-1)+P_big(t-1))*arg.gmax1; 	
	g2=P_small(t-1)/(arg.kG+P_small(t-1)+P_big(t-1))*arg.gmax2;
	g(t-1)=g1+g2;

	% fluxes
    PP_big(t)=u_big*P_big(t-1);
    PP_small(t)=u_small*P_small(t-1); 
    
	G1(t)=g1*Z(t-1); 
	G2(t)=g2*Z(t-1); 
	death_Z(t)=arg.mZ*Z(t-1)^2;
	excretion_Z(t)=arg.eZ*Z(t-1);
	regeneration_PO4(t)=arg.epsilon*excretion_Z(t);
    p_fec(t) = excretion_Z(t)-regeneration_PO4(t);
    
    max_available_PO4 = PO4(t-1)+Nsupply(t)*arg.dt+regeneration_PO4(t)*arg.dt;
	if PP_small(t)>max_available_PO4/arg.dt, PP_small(t)=max_available_PO4/arg.dt; end
    max_available_PO4=max_available_PO4-PP_small(t)*arg.dt;
	if PP_big(t)>max_available_PO4/arg.dt, PP_big(t)=max_available_PO4/arg.dt; end
    
	% carbon-based nutrients and biomass
	PO4(t)=PO4(t-1)+Nsupply(t)*arg.dt+regeneration_PO4(t)*arg.dt-PP_small(t)*arg.dt-PP_big(t)*arg.dt; PO4(PO4<=0)=0;
    
    P_big(t)=P_big(t-1)+PP_big(t)*arg.dt-G1(t)*arg.dt ;P_big(P_big<=0)=0;
	P_small(t)=P_small(t-1)+PP_small(t)*arg.dt-G2(t)*arg.dt; P_small(P_small<=0)=0;
    Z(t)=Z(t-1)+G1(t)*arg.dt+G2(t)*arg.dt-excretion_Z(t)*arg.dt-death_Z(t)*arg.dt; Z(Z<=0)=0;
    
    % export 
    Export(t) =p_fec(t)*arg.dt+death_Z(t)*arg.dt; 
    
    % Budget
    Budget(t) = PO4(t)*arg.dt-Export(t)*arg.dt; %Budget(Budget<=0)=0;

end

% save('outputs/maxPP_big1', 'maxPP_big1');
% save('outputs/maxPP_big2', 'maxPP_big2');
% save('outputs/maxPP_small1', 'maxPP_small1');
% save('outputs/maxPP_small2', 'maxPP_small2');

%% -------------- Ouputs

units=struct('time','days','Nsupply','mmolC m^{-3} d^{-1}',...
	'P_small','mmolC m^{-3}','P_big','mmolC m^{-3}','Z_small','mmolC m^{-3}','Z_big','mmolC m^{-3}',...
	'PO4','mmolC m^{-3}','Chl','mg m^{-3}','PP','gC m^{-3}/yr',...
	'u_small','d^{-1}','u_big','d^{-1}','g_small','d^{-1}','g_big','d^{-1}', 'PP_small', 'gC m^{-3}/yr', 'PP_big','gC m^{-3}/yr', 'Export', 'mmolC m^{-3} d^{-1}','Budget','mmolC m^{-3} d^{-1}','death_Z', 'mmolC m^{-3} d^{-1}', 'p_fec','mmolC m^{-3} d^{-1}');

output=struct('units',units,'time',time,'Nsupply',Nsupply,...
	'P_small',P_small,'P_big',P_big,'Z', Z,'PO4',PO4,...
	'Chl',P_small*12/arg.cChl_small+P_big*12./arg.cChl_big,'PP',(PP_big+PP_small)*1E-3*12*365.25,...
	'u_small',u_small,'u_big',u_big,'g',g, 'PP_small', PP_small,'PP_big', PP_big, 'Export', Export,'Budget', Budget,'death_Z',death_Z, 'p_fec', p_fec, 'attributs',struct('arg',arg));
varargout={output}; varargout=varargout(1:nargout);


% save('outputs/u_big','u_big')
% save('outputs/P_big','P_big')
% save('outputs/PP_big','PP_big')
%% -------------- Figures

if flag.plot

    %Temporal evolution
	figure, hold on
	plot(output.time,output.P_small,'LineWidth',2, 'Color','g')
	plot(output.time,output.P_big,'LineWidth',2,'Color', '#77AC30')
	plot(output.time,output.Z,'LineWidth',2,'Color','#00FFFF')
    plot(output.time,output.PO4,'LineWidth',2, 'Color','m')
	ylabel(output.units.P_small)
	if min(output.time)>datenum(1900,1,1), datetick('x','keeplimits'), xlabel('Time')
	else, xlabel('Time (days)')
	end
	legend({'P\_small','P\_big','Z','PO4'})
	title('Model output (plankon concentration over time)')
    xlim([min(output.time) max(output.time)]);
    ylim([0 1]);
    
    %Temporal evolution budget
	figure, hold on
	plot(output.time,output.Budget,'LineWidth',2, 'Color','k')
	if min(output.time)>datenum(1900,1,1), datetick('x','keeplimits'), xlabel('Time')
	else, xlabel('Time (days)')
	end
	legend({'Budget'})
	title('Model output (Budget over time)')
    xlim([min(output.time) max(output.time)]);
    %ylim([0 1]);

%     %Temporal evolution
% 	figure, hold on
% 	plot(output.time,output.u_small,'LineWidth',2)
% 	plot(output.time,output.u_big,'LineWidth',2)
% 	legend({'P\_small','P\_big'})
% 	title('Model output (plankon concentration over time)')
%     xlim([min(output.time) max(output.time)]);

    %Monod equation
    figure, hold on
    N_theo_PO4 = linspace(0,10,length(arg.time));
    f_monod2(N_theo_PO4,arg.umax_small,arg.kP_small);
    f_monod2(N_theo_PO4,arg.umax_big,arg.kP_big);
    f_monod2(output.PO4,arg.umax_small,arg.kP_small);
    f_monod2(output.PO4,arg.umax_big,arg.kP_big);
    xlabel('[PO4] mmolC.m³')
    ylabel('\mu d^{-1}');
    legend({'P\_small\_theo','P\_big\_theo','P\_small\_mod','P\_big\_mod' })
     
end

return

File: ga_model_2P1Z_v8.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Jan 18 11:23:36 2023

@author: loms
"""
"""
Implementation of the Growth Model 
                iP1Z
with growth rate and grazing rates = monod equation

i1 = 1 phytoplancton
i2 = 2 phytoplankton
                
Created on Tue Dec 20 12:39:07 2022

@author: loms
"""
import matplotlib.pyplot as plt
import numpy as npy
import numpy as np

plt.close('all')

# timestep
dt = 0.2
# the time at which the simulation ends (days)
end_time = 600
# creates a time vector from 0 to end_time, seperated by a timestep
time = npy.arange(0,end_time,dt)

"""" definition of model variables"""
##Creating empty vectors
# growth rates (d^-1)
u_1=[]
u_2=[]			
# grazing rates (d^-1)
g_1=[]
g_2=[]				
#Primary production
PP_1=[]
PP_2=[]
#Grazing
G1=[]
G2=[]
# biomass of Phytoplankton and Zooplankton (mmolC/m³)
P_1=[] 			
P_2=[] 						
Z = []
#Phosphate concentration
PO4 = []
Psupply=[]
#exc and regeneration
exc_Z=[]
PO4_reg=[]
p_fec=[]
#death
death_Z=[]
#export
Export=[]

# initial conditions at time=0	
P0_1 = 0.6
P0_2 = 0.1
Z0 = 0.6
PO40 = 0.5

"""" definition of model parameters"""

## Monod parameters
# growth rates (d^-1)
umax_1 = 1.9872
umax_2 = 2.7648 
# grazing rates (d^-1)
gmax_1 = 1.4226
gmax_2= 1.4226
# half saturation constants
kP_1 = 1
kP_2 = 2
kZ = 5

## Other model parameters
# Z exc fraction (d^-1)
eZ = 0.1
# fraction of Z exc that is available as regenerated PO4
epsilon = 0.75
# quadratic mortality of Z (mmolC^{-1} m^{-3} d^{-1})
mZ = 0.005

##Apport externe de nutriment 
Psupply_moy = 0.02

#Psupply constant
Psupply =[Psupply_moy] * len(time)

#Psupply variable

#Sinusoïdalement
# b = 0.05
# w = 1
# Psupply_sin = []
# for i in range(len(time)):
#     Psupply_sin.append(Psupply_moy+ b * np.sin(w*i))
#     Psupply = Psupply_sin 

#Pulsé
# T = 4
# b = (Psupply_moy*0.25*10)/(1-np.exp(-0.25*T))-Psupply_moy
# Psupply_sin = []
# for i in range(len(time)):
#     Psupply_sin.append(Psupply_moy+b*np.sin(w*i))
#     Psupply = Psupply_sin 



def ga_model_2P1Z_v8():
    
    ## -------------- Parameters
    global dt, end_time, time, umax_1, umax_2, gmax_1, gmax_2, kP_1, kP_2, KZ, eZ, epsilon, mZ, u_1, u_2, P_1, P_2, Z, u0_1, u0_2, P0_1, P0_2, Z0, Psupply
    
    # Initial conditions at time=0			
    P_1.append(P0_1)
    P_2.append(P0_2)
    Z.append(Z0)
    PO4.append(PO40)
    
    
    nb_time=len(time)
    
    for t in range(0,nb_time-1):
        # growth and grazing rates
        u_1next=PO4[t-1]/(kP_1+PO4[t-1])*umax_1
        u_2next=PO4[t-1]/(kP_2+PO4[t-1])*umax_2
       
        u_1.append(u_1next)
        u_2.append(u_2next)
       
        g_1next=P_1[t-1]/(kZ+P_1[t-1]+P_2[t-1])*gmax_1
        g_2next=P_2[t-1]/(kZ+P_1[t-1]+P_2[t-1])*gmax_2
       
        g_1.append(g_1next)
        g_2.append(g_2next)
   
   	    # fluxes
        PP_1next=u_1[t]*P_1[t-1]
        PP_2next=u_2[t]*P_2[t-1]
       
        PP_1.append(PP_1next)
        PP_2.append(PP_2next)
       
        G1_next=g_1[t]*Z[t-1]
        G2_next=g_2[t]*Z[t-1]
        death_Znext=mZ*Z[t-1]**2
        exc_Znext=eZ*Z[t-1]
       
        G1.append(G1_next)
        G2.append(G2_next)
        death_Z.append(death_Znext)
        exc_Z.append(exc_Znext)
       
        PO4_regnext=epsilon*exc_Z[t]
        p_fecnext=(1-epsilon)*exc_Z[t]
       
        PO4_reg.append(PO4_regnext)
        p_fec.append(p_fecnext)
       
       
        #PP_1 puise en premier dans le pool de nutriment
        max_available_PO4 = PO4[t-1]+Psupply[t]*dt+PO4_reg[t]*dt
        if PP_1[t]>max_available_PO4/dt:
            PP_1[t]=max_available_PO4/dt
            
        max_available_PO4=max_available_PO4-PP_1[t]*dt
        if PP_2[t]>max_available_PO4/dt:
            PP_2[t]=max_available_PO4/dt
        
        # export 
        Export_next =p_fec[t]+death_Z[t]
        Export.append(Export_next)
        
        # carbon-based nutrients and biomass
        PO4_next=PO4[t-1]+Psupply[t]*dt+PO4_reg[t]*dt-PP_1[t]*dt-PP_2[t]*dt
        if PO4_next<=0:
            PO4_next=0
            
        P_1_next=P_1[t-1]+PP_1[t]*dt-G1[t]*dt
        if P_1_next<=0:
            P_1_next=0
        
        P_2_next=P_2[t-1]+PP_2[t]*dt-G2[t]*dt
        if P_2_next<=0:
            P_2_next=0
            
        Z_next=Z[t-1]+G1[t]*dt+G2[t]*dt-exc_Z[t]*dt-death_Z[t]*dt
        if Z_next<=0:
            Z_next=0
        
        P_1.append(P_1_next)
        P_2.append(P_2_next)
        Z.append(Z_next)
        PO4.append(PO4_next)
    
    
    # Figures
    plt.figure(1)
    plt.plot(time, P_1, label='P_1', color="chartreuse")
    plt.plot(time, P_2, label='P_2',color="green")
    plt.plot(time, Z, label='Zoo', color="aqua")
    plt.plot(time,PO4, label='PO4',color="magenta")
    plt.xlabel('Time')
    plt.ylabel('mmolC.m^-3')
    plt.legend()
    plt.title('model outputs (concentration over time)')
    
    
    plt.figure(2)
    # Calcul des dérivées de P_1 et P_2 par rapport au temps
    dp1_dt = np.gradient(P_1, dt)
    dp2_dt = np.gradient(P_2, dt)
    
    # Tracé du portrait de phase
    plt.plot(P_2, P_1)
    plt.xlabel('P_2')
    plt.ylabel('P_1')
    
    # Tracé du champ de vecteur sur le portrait de phase
    plt.quiver(P_2, P_1, dp2_dt, dp1_dt)
    

    
    # plt.figure(2)
    # plt.plot(time,Export,label='export',color="black")
    # plt.plot(time,Psupply,label='Psupply',color="red")
    # plt.xlabel('Time (d)')
    # plt.ylabel('Flux (mmolC.m-3.d-1)')
    # plt.legend()
    # plt.title('Budget')
    
    # plt.figure(3)
    # plt.plot(u_1,P_1,label='P_1')
    # plt.plot(u_2,P_2, label='P_2')
    # plt.xlabel('Growth rate')
    # plt.ylabel('Biomass')
    # plt.legend()
    # plt.title('Behrenfeld model outputs')
    
    
    
#Call the function       
ga_model_2P1Z_v8()